sdk/eventhub/event-hubs/src/eventHubSender.ts

```diff
@@ -512,21 +501,137 @@ export class EventHubSender extends LinkEntity {
+          log.sender(
+            "[%s] Sender '%s', got event accepted.",
+            this._context.connectionId,
+            this.name
+          );
+          resolve();
+        };
+
+        const onAborted = () => {
+          removeListeners();
+          rejectOnAbort();
+        };
+
+        const onReleased = (context: EventContext) => {
+          removeListeners();
+          log.error("[%s] Sender '%s', got event released.", this._context.connectionId, this.name);
+          let err: Error;
+          if (context!.delivery!.remote_state!.error) {
+            err = translate(context!.delivery!.remote_state!.error);
+          } else {
+            err = new Error(
+              `[${this._context.connectionId}] Sender '${this.name}', ` +
+                `received a release disposition.Hence we are rejecting the promise.`
+            );
+          }
+          log.error(err);
+          reject(err);
+        };
+
+        const onModified = (context: EventContext) => {
+          removeListeners();
+          log.error("[%s] Sender '%s', got event modified.", this._context.connectionId, this.name);
+          let err: Error;
+          if (context!.delivery!.remote_state!.error) {
+            err = translate(context!.delivery!.remote_state!.error);
+          } else {
+            err = new Error(
+              `[${this._context.connectionId}] Sender "${this.name}", ` +
+                `received a modified disposition.Hence we are rejecting the promise.`
+            );
+          }
+          log.error(err);
+          reject(err);
+        };
+
+        const removeListeners = (): void => {
+          clearTimeout(waitTimer);
+          // When `removeListeners` is called on timeout, the sender might be closed and cleared
+          // So, check if it exists, before removing listeners from it.
+          if (abortSignal) {
+            abortSignal.removeEventListener("abort", onAborted);
+          }
+          if (this._sender) {
+            this._sender.removeListener(SenderEvents.rejected, onRejected);
+            this._sender.removeListener(SenderEvents.accepted, onAccepted);
+            this._sender.removeListener(SenderEvents.released, onReleased);
+            this._sender.removeListener(SenderEvents.modified, onModified);
+          }
+        };
+
+        const actionAfterTimeout = () => {
+          removeListeners();
+          const desc: string =
+            `[${this._context.connectionId}] Sender "${this.name}" with ` +
+            `address "${this.address}", was not able to send the message right now, due ` +
+            `to operation timeout.`;
+          log.error(desc);
+          const e: Error = {
+            name: "OperationTimeoutError",
+            message: desc
+          };
+          return reject(translate(e));
+        };
+
+        const waitTimer = setTimeout(
+          actionAfterTimeout,
+          getRetryAttemptTimeoutInMs(options.retryOptions)
+        );
```

Comment:
Due to the re-arrangement of the timer, if `if (abortSignal && abortSignal.aborted)` below returns true, then we end up returning a rejected promise (which is right) without clearing the timer.

Please consider keeping the previous order of the callbacks/helper-functions and only adding the link creation part at the right place

Reply:
The ordering in the recent commit of https://github.com/Azure/azure-sdk-for-js/pull/4319/commits/eaaf17e92e7532f69c79fe8d74fd30f3db6ae78b is better, but still needs work.

In the current setup, if the abort signal is fired when the async process of `init()` is in progress, then the promise is being rejected only after `init()` is complete. This is because `init()` is being called before the event handler for `abort` event was added.

Please make the below changes
- Add the event handler for abort before calling init() and before starting the timer
- Remove the event handler for abort before rejecting the promise when `init()` fails

Reply:
https://github.com/Azure/azure-sdk-for-js/pull/4322#discussion_r304612190 thread specifically clarifies about abort during init().

This scenario if we want to address would then need to apply for the managementRequest as well correct? Or did we specifically exclude this because of complexity involved? (In not having access to removeListeners() on sender link from the SDK?) It just feels like UX is not consistent in both cases.

Okay to implement the optimization and new corner cases from this.

Reply:
This scenario applies to managementRequest as well i.e. the scenario of abort signal being fired when the async process of init() is in progress.

I have updated the comment thread there. Please see https://github.com/Azure/azure-sdk-for-js/pull/4322/files#r305599979